src/doctopus/db.clj

```diff
@@ -0,0 +1,199 @@
+            [korma.core :refer :all]
+            [korma.db :refer [defdb postgres]]
+            [taoensso.timbre :as log]))
+
+(defn- now
+  []
+  "returns a string for 'now' in SQL Timestamp format"
+  (to-sql-time (clj-time/now)))
+
+(defn- add-updated
+  "adds updated field to a map with given time"
+  ([fields] (add-updated fields (now)))
+  ([fields now-string] (assoc fields :updated now-string)))
+
+(defn- ->kebab-keys
+  "converts all keys in a map to kebab-case keywords"
+  [fields]
+  (transform-keys ->kebab-case-keyword fields))
+
+(defn- ->snake-keys
+  "converts all keys in a map to snake-case keywords"
+  [fields]
+  (transform-keys ->snake_case_keyword fields))
+
+(defdb main (postgres
+             (select-keys (get-in (server-config) [:database :dev])
+                          [:db :user :password :host :port])))
+
+(declare head-tentacle-mappings)
+(defentity tentacles
+  (pk :name)
+  (has-many head-tentacle-mappings)
+  (prepare add-updated)
+  (prepare (fn [{html-commands :html-commands :as tentacle}]
+             (if html-commands
+               (assoc tentacle :html-commands (string/join "\n" html-commands))
+               tentacle)))
+  (prepare ->snake-keys)
+  (transform ->kebab-keys)
+  (transform (fn [{html-commands :html-commands :as tentacle}]
+               (if html-commands
+                 (assoc tentacle :html-commands (split-lines html-commands))
+                 tentacle))))
+
+
+(defentity heads
+  (pk :name)
+  (prepare add-updated)
+  (prepare ->snake-keys)
+  (transform ->kebab-keys)
+  (has-many head-tentacle-mappings))
+
+
+;; ### Head <-> Tentacle Join Table
+;; Heads and Tentacles can share the notorious SQL many-to-many
+;; relationship. Further, having either a head or a tentacle record _on the
+;; database_ storing information about the other gets conceptually weird. Enter:
+;; the join table.
+;;
+;; Join tables capture many-to-many relationships by inserting themselves in the
+;; middle, effectively creating a one-to-many relationship between itself and
+;; both of the other tables you're attempting to look up. Getting from Heads to
+;; Tentacles (or vice-versa) is now a set of left-joins. Easy peasy!
+(defentity head-tentacle-mappings
+  (table :head_tentacle_mappings)
+  (prepare ->snake-keys)
+  (transform ->kebab-keys)
+  (belongs-to heads)
+  (belongs-to tentacles))
+
+(defentity tentacles-by-head
+  (table (subselect head-tentacle-mappings
+                    (join tentacles (= :head_tentacle_mappings.tentacle_name :tentacles.name))
+                    (fields :head_tentacle_mappings.head_name
+                            :tentacles.name
+                            :tentacles.output_root
+                            :tentacles.html_commands
+                            :tentacles.source_control
+                            :tentacles.entry_point))
+         :tentacles_by_head)
+  ;; These don't seem to have any effect and I have no idea why.
```

Comment:
Yeah, that part is fucky. I'm still trying to figure out if I'm doing something wrong, or if it's a bug in Korma, or.... space.... goblins?